clickhouse-admin/src/http_entrypoints.rs

```diff
@@ -46,8 +49,26 @@ impl ClickhouseAdminServerApi for ClickhouseAdminServerImpl {
     ) -> Result<HttpResponseCreated<ReplicaConfig>, HttpError> {
         let ctx = rqctx.context();
         let replica_server = body.into_inner();
-        let output =
-            ctx.clickward().generate_server_config(replica_server.settings)?;
+        let current_generation = ctx.generation().await;
+        let incoming_generation = replica_server.generation();
+
+        // If the incoming generation number is lower, then we have a problem.
+        // We should return an error instead of silently skipping the configuration
+        // file generation.
+        if let Some(current) = current_generation {
+            if current > incoming_generation {
+                return Err(HttpError::for_internal_error(
+                    "current generation is greater than incoming generation"
+                        .to_string(),
+                ));
+            }
+        };
+
+        let output = ctx.clickward().generate_server_config(replica_server)?;
+
+        // We want to update the generation number only if the config file has been
+        // generated successfully.
+        *ctx.generation.lock().await = Some(incoming_generation);
```

Comment:
Is there a TOCTOU problem here, in that `ctx.generation()` could have changed between when we checked it above and when we reacquire the lock here to set it?

Reply:
Hm, I guess that depends on how reconfigurator works? How often is the generation changing? 

I decided to update the generation number once the config file had been successfully generated, because if it hadn't, then the zone wouldn't be fully in that generation. Do you think it makes more sense to update the generation immediately?

Reply:
I think I'd consider this outside the context of reconfigurator. If this endpoint is called multiple times concurrently with different incoming generations, does it behave correctly? That way we don't have an implicit dependency between the correctness of this endpoint and the behavior or timing of reconfigurator.

Sorry for the dumb questions, but - is it safe for two instances of `generate_server_config()` to be running concurrently? I think that has implications on what we need to do with the lock on `generation`.

Reply:
> I think I'd consider this outside the context of reconfigurator. If this endpoint is called multiple times concurrently with different incoming generations, does it behave correctly? 

I guess there could be an error if two `generate_server_config()`s with different generation numbers are running
and they both read an initial value for generation, but one with the lower number manages to write after the one with the higher one.

Thanks for the input! I guess that settles it, I'll update the number immediately after reading. I was on the fence about this one anyway. Even if the config is borked, it'll be borked in that generation

Reply:
Thanks for the detailed answer!

> Hm, I'm not sure that's enough. We may need to write the config file while holding the lock too, I think?

Yep, that makes total sense

Reply:
Yeah, you are right @jgallagher. These requests all need to be serialized. (I know you are currently writing up some options, just wanted to drop a note).